File: app/src/main/java/com/kararnab/contacts/room/ContactRepository.java
package com.kararnab.contacts.room;

import android.app.Application;
import android.arch.lifecycle.LiveData;
import android.os.AsyncTask;

import java.util.List;

/**
 * Abstracted Repository as promoted by the Architecture Guide.
 * https://developer.android.com/topic/libraries/architecture/guide.html
 */

public class ContactRepository {

    private ContactDao mContactDao;
    private LiveData<List<Contact>> mAllContacts;

    // Note that in order to unit test the WordRepository, you have to remove the Application
    // dependency. This adds complexity and much more code, and this sample is not about testing.
    // See the BasicSample in the android-architecture-components repository at
    // https://github.com/googlesamples
    ContactRepository(Application application) {
        ContactRoomDatabase db = ContactRoomDatabase.getDatabase(application);
        mContactDao = db.contactDao();
        mAllContacts = mContactDao.getAlphabetizedWords();
    }

    // Room executes all queries on a separate thread.
    // Observed LiveData will notify the observer when the data has changed.
    LiveData<List<Contact>> getAllContacts() {
        return mAllContacts;
    }

    // You must call this on a non-UI thread or your app will crash.
    // Like this, Room ensures that you're not doing any long running operations on the main
    // thread, blocking the UI.
    public void insert(Contact contact) {
        new insertAsyncTask(mContactDao).execute(contact);
    }

    private static class insertAsyncTask extends AsyncTask<Contact, Void, Void> {

        private ContactDao mAsyncTaskDao;

        insertAsyncTask(ContactDao dao) {
            mAsyncTaskDao = dao;
        }

        @Override
        protected Void doInBackground(final Contact... params) {
            mAsyncTaskDao.insert(params[0]);
            return null;
        }
    }
}

File: app/src/main/java/com/kararnab/contacts/room/ContactRepository.kt
package com.kararnab.contacts.room

import android.app.Application
import androidx.lifecycle.LiveData

/**
 * Abstracted Repository as promoted by the Architecture Guide.
 * https://developer.android.com/topic/libraries/architecture/guide.html
 */
class ContactRepository(application: Application) {
    private val mContactDao: ContactDao

    // Room executes all queries on a separate thread.
    // Observed LiveData will notify the observer when the data has changed.
    val allContacts: LiveData<List<Contact>>

    // You must call this on a non-UI thread or your app will crash.
    // Like this, Room ensures that you're not doing any long running operations on the main
    // thread, blocking the UI.
    suspend fun insert(contact: Contact) {
        mContactDao.insert(contact)
    }

    suspend fun update(contact: Contact) {
        mContactDao.update(contact)
    }

    fun getContact(id: Int, fallbackContact: Contact) : Contact {
        val contacts = mContactDao.getContact(id)
        return if(contacts.isNotEmpty()) {
            contacts[0]
        }else{
            fallbackContact
        }
    }

    /*private class insertAsyncTask internal constructor(private val mAsyncTaskDao: ContactDao) : AsyncTask<Contact?, Void?, Void?>() {
        protected override fun doInBackground(vararg params: Contact): Void? {
            mAsyncTaskDao.insert(params[0])
            return null
        }
    }*/

    // Note that in order to unit test the WordRepository, you have to remove the Application
    // dependency. This adds complexity and much more code, and this sample is not about testing.
    // See the BasicSample in the android-architecture-components repository at
    // https://github.com/googlesamples
    init {
        val db = ContactRoomDatabase.getInstance(application)
        mContactDao = db.contactDao()
        allContacts = mContactDao.getAlphabetizedWords()
    }
}